cns/restserver/ipam.go

```diff
@@ -92,40 +92,58 @@ func (service *HTTPRestService) releaseIPConfigHandler(w http.ResponseWriter, r
-	// Ensure PendingProgramming IPs will be release before Available ones.
-	ipStateTypes := [2]string{cns.PendingProgramming, cns.Available}
-
+// MarkIPAsPendingRelease will set the IPs which are in PendingProgramming or Available to PendingRelease state
+// It will try to update [totalIpsToRelease]  number of ips.
+func (service *HTTPRestService) MarkIPAsPendingRelease(totalIpsToRelease int) (map[string]cns.IPConfigurationStatus, error) {
+	pendingReleasedIps := make(map[string]cns.IPConfigurationStatus)
 	service.Lock()
 	defer service.Unlock()
-	for _, ipStateType := range ipStateTypes {
-		pendingReleaseIPs := service.markSpecificIPTypeAsPending(numberToMark, ipStateType)
-		for uuid, pependingReleaseIP := range pendingReleaseIPs {
-			allReleasedIPs[uuid] = pependingReleaseIP
-		}
-		numberToMark -= len(pendingReleaseIPs)
-		if numberToMark == 0 {
-			return allReleasedIPs, nil
-		}
+
+	var count int
+	for uuid, ipcConfig := range service.PodIPConfigState {
+        if ipcConfig.State == cns.PendingProgramming {
+            if err := service.updateIPConfigState(uuid, cns.PendingRelease); err != nil {
+                return nil, err
+            }
+
+            pendingReleasedIps[uuid] = ipcConfig
+			count++
+			
+			if count == totalIpsToRelease {
+				return pendingReleasedIps, nil
+			}	
+    	} 
+	}
+	
+	// if not all expected IPs are set to PendingRelease, then check the Available IPs 
+	for uuid, ipcConfig := range service.PodIPConfigState {
+        if ipcConfig.State == cns.Available {
+            if err := service.updateIPConfigState(uuid, cns.PendingRelease); err != nil {
+                return nil, err
+            }
+
+            pendingReleasedIps[uuid] = ipcConfig
+			count++
+			
+			if count == totalIpsToRelease {
+				return pendingReleasedIps, nil
+			}	
+    	} 
 	}
-	return nil, fmt.Errorf("Failed to mark %d IP's as pending, only marked %d IP's", numberToMark, len(allReleasedIPs))
+
+	logger.Printf("[MarkIPAsPendingRelease] Set total ips to PendingRelease %d, expected %d", count, totalIpsToRelease)
+	return pendingReleasedIps, nil
 }
 
-func (service *HTTPRestService) markSpecificIPTypeAsPending(numberToMark int, ipStateType string) map[string]cns.IPConfigurationStatus {
-	pendingReleaseIPs := make(map[string]cns.IPConfigurationStatus)
-	for uuid, mutableIPConfig := range service.PodIPConfigState {
-		if mutableIPConfig.State == ipStateType {
-			mutableIPConfig.State = cns.PendingRelease
-			service.PodIPConfigState[uuid] = mutableIPConfig
-			pendingReleaseIPs[uuid] = mutableIPConfig
-			if len(pendingReleaseIPs) == numberToMark {
-				return pendingReleaseIPs
-			}
-		}
-	}
-	return pendingReleaseIPs
+func (service *HTTPRestService) updateIPConfigState(ipId string, updatedState string) error {
+	if ipConfig, found := service.PodIPConfigState[ipId]; found {
+		logger.Printf("[updateIPConfigState] Changing existingIpConfig [%v] state [%s] to [%s]", ipConfig, ipConfig.State, updatedState)
```

Comment:
This might be difficult to read since the entire ip config struct is being printed at the beginning. It's also not printed with fields which makes it harder to know what value corresponds to what. If you still want to dump the entire struct and not the identifier, I'd recommend something like
```
logger.Printf("[updateIPConfigState] Changing ip id [%s] state to [%s]. Current config: %+v", ipId, updatedState, ipConfig)
```
which will also print the struct fields along with the data